### PR TITLE
use TDNFJoin() to construct paths

### PR DESCRIPTION
**Pull request by oliverkurth**
This adds the `TDNFJoin()` function to create directory paths.

### CODE REVIEW

File: common/utils.c [hunk 1]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+    int nAbsolute = 0;
+
+    if (!ppszPath)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    va_start(ap, ppszPath);
+    for(pszNode = va_arg(ap, char *); pszNode; pszNode = va_arg(ap, char *))
```
Comment (sshedi) [inline common/utils.c:808]: if this loop is for counting number of va args, then it doesn't work.
get rid of next for loop (at line 795) and process args one by one withing this loop.

File: common/utils.c [hunk 2]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for(pszNode = va_arg(ap, char *), i = 0; pszNode; pszNode = va_arg(ap, char *), i++)
+    {
+        dwError = TDNFAllocateString(pszNode, &pszNodeCopy);
+        BAIL_ON_TDNF_ERROR(dwError);
+        pszNodeTmp = pszNodeCopy;
+        if (i == 0 && *pszNodeTmp == '/')
+        {
+            nAbsolute = 1;
```
Comment (sshedi) [inline common/utils.c]: nit: do pszNodeTmp++ here, saves one iteration :)

File: common/utils.c [hunk 3]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+        }
+        while(*pszNodeTmp == '/') pszNodeTmp++;
```
Comment (sshedi) [inline common/utils.c:837]: User strspn instead.

File: common/utils.c [hunk 4]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+        pszTmp = pszNodeTmp + strlen(pszNodeTmp) - 1;
```
Comment (sshedi) [inline common/utils.c]: store `strlen(pszNodeTmp)` in a variable. it's again used in line 813

File: common/utils.c [hunk 5]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+
+    pszTmp = pszResult;
+    if (nAbsolute)
+    {
+        *pszTmp = '/';
+        pszTmp++;
+    }
+    for (i = 0; i < nCount; i++)
+    {
+        strcpy(pszTmp, ppszNodes[i]);
```
Comment (sshedi) [inline common/utils.c]: Use strcat here?

File: common/utils.c [hunk 6]
```diff
@@ -760,3 +760,89 @@ TDNFStringMatchesOneOf(const char *pszSearch, char **ppszList, int *pRet)
...
+        while(*pszNodeTmp == '/') pszNodeTmp++;
+        pszTmp = pszNodeTmp + strlen(pszNodeTmp) - 1;
+        while(pszTmp >= pszNodeTmp && *pszTmp == '/')
+        {
+            *pszTmp = 0;
+            pszTmp--;
+        }
+        dwError = TDNFAllocateString(pszNodeTmp, &ppszNodes[i]);
+        BAIL_ON_TDNF_ERROR(dwError);
+        nLength += strlen(pszNodeTmp) + 1;
```
Comment (sshedi) [inline common/utils.c]: you can just keep appending path to a resultant buffer upon every iteration and reallocate memory as needed.

**Review by sshedi — CHANGES_REQUESTED** [review]

File: common/utils.c (hunk 2 above)
Comment (oliverkurth) [reply common/utils.c]: Thx - even better, I need the `while` loop only if the condition is true.

File: common/utils.c (hunk 3 above)
Comment (oliverkurth) [reply common/utils.c:837]: That's not making the code any more readable or efficient.

File: common/utils.c (hunk 4 above)
Comment (oliverkurth) [reply common/utils.c]: Thx! Done.

File: common/utils.c (hunk 1 above)
Comment (oliverkurth) [reply common/utils.c:808]: Last arg must be `NULL`. As discussed this internally.

File: common/utils.c (hunk 5 above)
Comment (oliverkurth) [reply common/utils.c]: Done.

File: common/utils.c (hunk 6 above)
Comment (oliverkurth) [reply common/utils.c]: Done.

**Review by sshedi — APPROVED** [review]
LGTM. One improvement I can think of is, if given an absolute path call NormalizePath function and use the return value as is.

**Comment by oliverkurth** [comment]
Build failure is unrelated: Photon 4.0 updated to openssl 3.0.0. Possible fix:
```
diff --git a/client/remoterepo.c b/client/remoterepo.c
index 22c870e..e53d0d6 100644
--- a/client/remoterepo.c
+++ b/client/remoterepo.c
@@ -170,9 +170,13 @@ TDNFGetDigestForFile(
     {
         pr_err("Digest Init Failed\n");
         dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
-        /*MD5 is not approved in FIPS mode. So, overrriding
-          the dwError to show the right error to the user */
+        /* MD5 is not approved in FIPS mode. So, overrriding
+           the dwError to show the right error to the user */
+#if defined(OPENSSL_VERSION_MAJOR) && (OPENSSL_VERSION_MAJOR >= 3)
+        if (EVP_default_properties_is_fips_enabled(NULL) && !strcasecmp(hash->hash_type, "md5"))
+#else
         if (FIPS_mode() && !strcasecmp(hash->hash_type, "md5"))
+#endif
         {
             dwError = ERROR_TDNF_FIPS_MODE_FORBIDDEN;
         }
```

**Pull request merged**